core/src/main/java/org/apache/iceberg/io/datafile/DataFileServiceRegistry.java

```diff
@@ -0,0 +1,420 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io.datafile;
+
+import java.util.Map;
+import java.util.ServiceLoader;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Registry which maintains the available {@link ReaderService} and {@link WriterService}
+ * implementations. Based on the `file format`, the required `data type` and the reader/writer
+ * `builderType` the registry returns the correct reader and writer service implementations. These
+ * services could be used to generate the correct reader and writer builders.
+ */
+public final class DataFileServiceRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(DataFileServiceRegistry.class);
+  private static final Map<Key, ReaderService> READ_BUILDERS = Maps.newConcurrentMap();
+  private static final Map<Key, WriterService<?>> WRITE_BUILDERS = Maps.newConcurrentMap();
+
+  static {
+    for (ReaderService service : ServiceLoader.load(ReaderService.class)) {
+      if (READ_BUILDERS.containsKey(service.key())) {
```

Comment:
From a JVM/class-loading POV it's better to move this static initializer and the two CHMs above into a separate static inner class, so that it's not triggered immediately when the `DataFileServiceRegistry` class is initialized.

Reply:
Do you have concrete examples when this change could help?
AFAIK restrictions on JVM classloaders are very limited wrt deciding when to load specific classes. I'm not sure moving the static initialization part to a static inner class could/would change a lot, but I can be mistaken.

Reply:
Added this change.
Thanks for the suggestion @snazy!